README.md
# exceedProb
This R package computes exceedance probabilities and associated confidence intervals. Currently supports general linear models, with a beta function for Cox models. Please see Segal (2019) for more information.

## Installation

```{r}
install.packages("exceedProb")
```

## Examples

```{r}
library(exceedProb)

# Sample mean -----------------------------------------------------------------
n <- 100
x <- rnorm(n = n)

theta_hat <- mean(x)
sd_hat <- sd(x)

cutoff <- seq(from = theta_hat - 0.5, to = theta_hat + 0.5, by = 0.1)

exceedProb(cutoff = cutoff, 
           theta_hat = theta_hat, 
           sd_hat = sd_hat, 
           alpha = 0.05, 
           d = 1,
           n = n,
           m = n)

# Linear regression -----------------------------------------------------------
n <- 100
beta <- c(1, 2)
x <-runif(n = n, min = 0, max = 10)
y <- rnorm(n = n, mean = cbind(1, x) %*% beta, sd = 1)

j <- 2
fit <- lm(y ~ x)
theta_hat <- coef(fit)[j]
sd_hat <- sqrt(n * vcov(fit)[j, j])

cutoff <- seq(from = theta_hat - 0.5, to = theta_hat + 0.5, by = 0.1)

exceedProb(cutoff = cutoff, 
           theta_hat = theta_hat, 
           sd_hat = sd_hat, 
           alpha = 0.05, 
           d = length(beta),
           n = n,
           m = n)
```

## References

Segal, B. D. (2019). Toward replicability with confidence intervals for the exceedance probability. The American Statistician. [doi:10.1080/00031305.2019.1678521](https://www.tandfonline.com/eprint/SUPYJD7NHZMHMNNH5VHU/full?target=10.1080/00031305.2019.1678521).


NEWS.md
# exceedProb 0.0.2

This versions of exceedProb includes the new function exceedProbCoxBoot: a beta function for computing bootstrap confidence intervals for the exceedance probability of Cox model parameters.


R/ep.R

# functions -------------------------------------------------------------------
tRoot <- function(delta, test_stat, df, conf_level) {
  #' This function is used to find the root for a t-distribution pivotal quantity 
  #'
  #' This function returns the difference between the lower tail
  #' probability of a non-central t-distribution and a confidence level q
  #' where the t-distribution has df degrees of freedom and 
  #' non-centrality parameter delta.
  #' @param delta Non-centrality parameter
  #' @param test_stat Test statistic at which to evaluate the t-distribution
  #' @param df Degrees of freedom
  #' @param conf_level Confidence level (usually alpha/2 or 1-alpha/2)
  #' @return dif Difference between t-distribution quantile and confidence level
  #' @export

  dif <- pnct(test_stat, df, delta) -  conf_level

  return(dif)
}

getDeltaCI <- function(test_stat, 
                  alpha, 
                  d,
                  n, 
                  interval) {
  #' Confidence intervals for noncentrality parameter of t-distribution
  #'
  #' This function obtains confidence intervals for the non-centrality
  #' parameter of a t-distribution.
  #' @param test_stat Test statistics
  #' @param alpha Significance level
  #' @param d Number of parameters in general linear model
  #' @param n Number of observations in initial study
  #' @param interval Interval within which to search for roots
  #' @return ep Exceedance probability with confidence intervals (vector if cutoff is scalar and matrix otherwise)
  #' @export

  delta_lower <- uniroot(f = tRoot, 
                         interval = interval, 
                         test_stat = test_stat, 
                         df = n - d,
                         conf_level = 1 - alpha / 2)$root

  delta_upper <- uniroot(f = tRoot, 
                         interval = interval, 
                         test_stat = test_stat, 
                         df = n - d,
                         conf_level = alpha / 2)$root

  delta_ci <- c(delta_lower, delta_upper)
  names(delta_ci) <- c("lower", "upper")
  return(delta_ci)
}

getDeltaCI <- Vectorize(getDeltaCI, vectorize.args = "test_stat")


exceedProb <- function(cutoff, 
                  theta_hat, 
                  sd_hat, 
                  alpha, 
                  d,
                  n, 
                  m,
                  interval = c(-100, 100),
                  lower_tail = FALSE) {
  #' Confidence intervals for the exceedance probability
  #'
  #' This function obtains confidence intervals for exceedance probability
  #' @param cutoff Cutoff values (scalar or vector)
  #' @param theta_hat Point estimate for the parameter of interest
  #' @param sd_hat Estimated standard deviation for the parameter of interest (Note: not the standard error)
  #' @param d Number of parameters in the general linear model
  #' @param alpha Significance level
  #' @param n Number of observations in the initial study
  #' @param m Number of observations in the replication study
  #' @param interval Interval within which to search for roots
  #' @param lower_tail If TRUE, reports lower tail probabilities
  #' @return ep Exceedance probability with confidence intervals
  #' @export
  #' @examples
  #' library(exceedProb)
  #' 
  #' # Sample mean -------------------------------------------------------
  #' n <- 100
  #' x <- rnorm(n = n)
  #' 
  #' theta_hat <- mean(x)
  #' sd_hat <- sd(x)
  #' 
  #' cutoff <- seq(from = theta_hat - 0.5, to = theta_hat + 0.5, by = 0.1)
  #' 
  #' exceedProb(cutoff = cutoff, 
  #'            theta_hat = theta_hat, 
  #'            sd_hat = sd_hat, 
  #'            alpha = 0.05, 
  #'            d = 1,
  #'            n = n,
  #'            m = n)
  #' 
  #' # Linear regression -------------------------------------------------
  #' n <- 100
  #' beta <- c(1, 2)
  #' x <-runif(n = n, min = 0, max = 10)
  #' y <- rnorm(n = n, mean = cbind(1, x) %*% beta, sd = 1)
  #' 
  #' j <- 2
  #' fit <- lm(y ~ x)
  #' theta_hat <- coef(fit)[j]
  #' sd_hat <- sqrt(n * vcov(fit)[j, j])
  #' 
  #' cutoff <- seq(from = theta_hat - 0.5, to = theta_hat + 0.5, by = 0.1)
  #' 
  #' exceedProb(cutoff = cutoff, 
  #'            theta_hat = theta_hat, 
  #'            sd_hat = sd_hat, 
  #'            alpha = 0.05, 
  #'            d = length(beta),
  #'            n = n,
  #'            m = n)

  if (length(cutoff) < 1 | !is.numeric(cutoff)) {
    stop("cutoff must be numeric a numeric vector with at least 1 element")
  }

  if (length(theta_hat) != 1 | !is.numeric(theta_hat)) {
    stop("theta_hat must be numeric scalar")
  }

  if (length(sd_hat) != 1 | !is.numeric(sd_hat)) {
    stop("sd_hat must be numeric scalar")
  }

  if (length(d) != 1 | !is.numeric(d)) {
    stop("d must be numeric scalar")
  }

  if (length(n) != 1 | !is.numeric(n)) {
    stop("n must be numeric scalar")
  }

  if (length(m) != 1 | !is.numeric(m)) {
    stop("m must be numeric scalar")
  }

  if (length(interval) != 2 | !is.numeric(interval)) {
    stop("interval must be a numeric vector of length 2")
  }

  if (diff(interval) <= 0) {
    stop("interval must have the form (a, b) for a < b")
  }

  if(!is.logical(lower_tail) | length(lower_tail) != 1) {
    stop("lower_tail must be a logical scalar")
  }

  test_stat <- sqrt(n) * (cutoff - theta_hat) / sd_hat

  delta_ci <- getDeltaCI(test_stat = test_stat,
                         d = d,
                         alpha = alpha, 
                         n = n,
                         interval = interval)

  if (lower_tail) {
    lower <- stats::pnorm(sqrt(m/n) * delta_ci["lower", ], lower.tail = TRUE)
    upper <- stats::pnorm(sqrt(m/n) * delta_ci["upper", ], lower.tail = TRUE)
    point <- stats::pnorm(q = sqrt(m) * (cutoff - theta_hat) / sd_hat, lower.tail = TRUE)
  } else {
    lower <- stats::pnorm(sqrt(m/n) * delta_ci["upper", ], lower.tail = FALSE)
    upper <- stats::pnorm(sqrt(m/n) * delta_ci["lower", ], lower.tail = FALSE)
    point <- stats::pnorm(q = sqrt(m) * (cutoff - theta_hat) / sd_hat, lower.tail = FALSE)
  }

  ep <- data.frame(cutoff = cutoff, point = point, lower = lower, upper = upper)
  rownames(ep) <- NULL

  return(ep)
}


R/ep_boot_cox.R

fitCox <- function(data, surv_formula, j, cutoff, n, m, lower_tail) {
  #' Point estimate of the exceedance probability for Cox model parameters
  #' 
  #' This function obtains point estimates for the exceedance probability for Cox model parameters.
  #' @param data survival data (data.frame)
  #' @param surv_formula Survival formula (character), e.g. "Surv(time, event) ~ group"
  #' @param j  Index of parameter for which the exceedance probability is obtained
  #' @param cutoff Cutoff values (scalar or vector)
  #' @param n Number of observations in the original study
  #' @param m Number of observations in the replication study (defaults to n if NULL)
  #' @param lower_tail If TRUE, reports lower tail probabilities
  #' @return point (scalar or vector) Point estimate of exceedance probability
  #' @export

  fit0 <- survival::coxph(stats::as.formula(paste0("survival::", surv_formula)), data = data)
  theta_hat <- stats::coef(fit0)[j]
  sd_hat <- sqrt(n * stats::vcov(fit0)[j, j])

  point <- stats::pnorm(q = sqrt(m) * (cutoff - theta_hat) / sd_hat, lower.tail = lower_tail)

  return(point)
}

exceedProbCoxBoot <- function(data, surv_formula, j, cutoff = NULL, m = NULL, alpha = 0.05, B = 1000, lower_tail = FALSE) {
  #' Bootstrap confidence intervals for the exceedance probability of Cox model parameters
  #' 
  #' This function obtains nonparametric bootstrap percentile confidence for Cox model parameters.
  #' Beta version.
  #' @param data survival data (data.frame)
  #' @param surv_formula Survival formula (character), e.g. "Surv(time, event) ~ group"
  #' @param j Index of parameter for which the exceedance probability is obtained
  #' @param cutoff Cutoff values (scalar or vector if supplied, otherwise set to +/- 0.5 of theta_hat)
  #' @param alpha Significance level
  #' @param m Number of observations in the replication study (defaults to n if NULL)
  #' @param B Number of bootstrap resamples
  #' @param lower_tail If TRUE, reports lower tail probabilities
  #' @return ep Exceedance probability with confidence intervals
  #' @export
  #' @examples
  #' library(exceedProb)
  #' 
  #' # Cox model -------------------------------------------------------
  #'
  #' # Simulate exponential data
  #' n <- 100
  #' m <- 100
  #' baseline_hazard <- 1
  #' theta <- 0.4
  #' p_censor <- 0.3
  #' prop_tx <- 0.5
  #'
  #' tx_indicator = rbinom(n = n, size = 1, prob = prop_tx)
  #' event_rate <- baseline_hazard * exp(theta * tx_indicator)
  #' censor_rate <- event_rate * p_censor / (1 - p_censor)
  #'
  #' event_time <- rexp(n = n, rate = event_rate)
  #' censor_time <- rexp(n = n, rate = censor_rate)
  #'
  #' time <- pmin(event_time, censor_time)
  #' event <- time == event_time
  #'
  #' surv_data = data.frame(time = time, event = event, group = tx_indicator)
  #'
  #' # Get exceedance probability with nonparametric bootstrap percentile confidence intervals
  #' ep <- exceedProbCoxBoot(data = surv_data,
  #'                         surv_formula = "Surv(time, event) ~ group",
  #'                         j = 1,
  #'                         B = 1000)

  # TODO(bsegal): Add checks to ensure input is valid

  n <- nrow(data)

  if(is.null(m)) {
    m <- n
  }

  fit0 <- survival::coxph(stats::as.formula(paste0("survival::", surv_formula)), data = data)
  theta_hat <- stats::coef(fit0)[j]
  sd_hat <- sqrt(n * stats::vcov(fit0)[j, j])

  if(is.null(cutoff)) {
    cutoff <- seq(from = theta_hat - 0.5, to = theta_hat + 0.5, by = 0.1)
  }

  point <- stats::pnorm(q = sqrt(m) * (cutoff - theta_hat) / sd_hat, lower.tail = lower_tail)

  boot_out <- matrix(NA, ncol = length(cutoff), nrow = B)

  for (b in 1:B) {
    data_star <- data[sample(n, size = n, replace = TRUE), ]
    boot_out[b, ] <- fitCox(data = data_star,
                            surv_formula = surv_formula,
                            j = j,
                            cutoff = cutoff,
                            n = n,
                            m = m,
                            lower_tail = lower_tail)
  }

  ci <- apply(boot_out, 2, stats::quantile, probs = c(alpha / 2, 1 - alpha / 2))

  ep <- data.frame(cutoff = cutoff, point = point, lower = ci[1, ], upper = ci[2, ])
  rownames(ep) <- NULL

  return(ep)
}


R/ep_cox_boot.R

CoxEP <- function(data, surv_formula, j, cutoff, n, m, lower_tail) {
  #' Point estimate of the exceedance probability for Cox model parameters
  #' 
  #' This function obtains point estimates for the exceedance probability for Cox model parameters.
  #' @param data survival data (data.frame)
  #' @param surv_formula Survival formula
  #' @param j Index of parameter for which the exceedance probability is obtained
  #' @param cutoff Cutoff values (scalar or vector)
  #' @param n Number of observations in the original study
  #' @param m Number of observations in the replication study (defaults to n if NULL)
  #' @param lower_tail If TRUE, reports lower tail probabilities
  #' @return point (scalar or vector) Point estimate of exceedance probability
  #' @export

  fit0 <- survival::coxph(surv_formula, data = data)
  theta_hat <- stats::coef(fit0)[j]
  sd_hat <- sqrt(n * stats::vcov(fit0)[j, j])

  point <- stats::pnorm(q = sqrt(m) * (cutoff - theta_hat) / sd_hat, lower.tail = lower_tail)

  return(point)
}

exceedProbCoxBoot <- function(data,
                              cox_fit,
                              j, 
                              alpha, 
                              R, 
                              cutoff = NULL, 
                              m = NULL, 
                              lower_tail = FALSE, 
                              sim = "model") {
  #' Bootstrap confidence intervals for the exceedance probability of Cox model parameters
  #' 
  #' This function obtains nonparametric bootstrap percentile confidence for Cox model parameters.
  #' Beta version.
  #' @param data survival data (data.frame)
  #' @param cox_fit (coxph.object) A fitted Cox model
  #' @param j Index of parameter for which the exceedance probability is obtained
  #' @param alpha Significance level
  #' @param R Number of bootstrap resamples
  #' @param cutoff Cutoff values (scalar or vector if supplied, otherwise set to +/- 0.5 of theta_hat)
  #' @param m Number of observations in the replication study (defaults to n if NULL)
  #' @param lower_tail If TRUE, reports lower tail probabilities; otherwise reports upper tail probabilities
  #' @param sim type of simulation, input to boot::censboot
  #' @return ep Exceedance probability with confidence intervals
  #' @export
  #' @examples
  #' library(exceedProb)
  #' library(survival)
  #' 
  #' # Cox model -------------------------------------------------------
  #'
  #' # Simulate exponential data
  #' n <- 50
  #' baseline_hazard <- 1
  #' theta <- 0.4
  #' p_censor <- 0.3
  #' prop_tx <- 0.5
  #' 
  #' tx_indicator = rbinom(n = n, size = 1, prob = prop_tx)
  #' event_rate <- baseline_hazard * exp(theta * tx_indicator)
  #' censor_rate <- event_rate * p_censor / (1 - p_censor)
  #' 
  #' event_time <- rexp(n = n, rate = event_rate)
  #' censor_time <- rexp(n = n, rate = censor_rate)

  #' time <- pmin(event_time, censor_time)
  #' event <- time == event_time
  #' surv_data = data.frame(time = time, event = event, group = tx_indicator)
  #'
  #' # Fit Cox model and get bootstrap percentile confidence intervals for the exceedance probability 
  #' # with model-based resampling (see documentation for boot::censboot)
  #' cox_fit <- coxph(Surv(time, event) ~ group, data = surv_data)
  #' ep <- exceedProbCoxBoot(data = surv_data,
  #'                         cox_fit = cox_fit,
  #'                         j = 1, 
  #'                         alpha = 0.05, 
  #'                         R = 500)
  #'
  #' # Plot results
  #' with(ep, plot(cutoff, point, type = "l"))
  #' with(ep, lines(cutoff, lower, lty = 2))
  #' with(ep, lines(cutoff, upper, lty = 2))

  event_chr <- as.character(cox_fit$formula[[2]])[3]

  # Todo(bsegal): revise code below so this warning is not necessary
  if(!is.logical(eval(parse(text = event_chr), data))) {
    stop(paste0("The event indicator must evaluate to a logical; ", event_chr, " does not."))
  }

  # TODO(bsegal): Add other checks to ensure input is valid

  km_surv <- survival::survfit(cox_fit)

  # pull out elements of model formula
  time_chr <- as.character(cox_fit$formula[[2]])[2]
  covariate_chr <- as.character(cox_fit$formula[[3]])

  # create formula for reverse KM
  time_revised_chr <- paste0(time_chr, " - 0.001*(", event_chr, ")")
  event_reverse_chr <- paste0("!(", event_chr, ")")
  rev_km_formula <- stats::as.formula(paste0("Surv(", time_revised_chr, ", ", event_reverse_chr, ") ~ ", covariate_chr))

  # assuming censoring distribution explained by same covariates as survival distribution
  km_cens <- survival::survfit(rev_km_formula, data = data)

  # using number of rows to scale variance,
  # assuming number of events is proportional to number of rows
  n <- nrow(data)

  if (is.null(m)) {
    m <- n
  }

  theta_hat <- stats::coef(cox_fit)[j]
  sd_hat <- sqrt(n * stats::vcov(cox_fit)[j, j])

  if(is.null(cutoff)) {
    cutoff <- seq(from = theta_hat - 1, to = theta_hat + 1, by = 0.1)
  }

  point <- stats::pnorm(q = sqrt(m) * (cutoff - theta_hat) / sd_hat, lower.tail = lower_tail)

  boot_out <- boot::censboot(data = data,
                             statistic = CoxEP,
                             R = R,
                             F.surv = km_surv,
                             G.surv = km_cens,
                             sim = sim,
                             cox = cox_fit,
                             surv_formula = cox_fit$formula,
                             j = j,
                             cutoff = cutoff,
                             n = n,
                             m = m,
                             lower_tail = lower_tail)

  ci <- apply(boot_out$t, 2, stats::quantile, probs = c(1 - alpha / 2, alpha / 2), type = 6)
  ep <- data.frame(cutoff = cutoff, point = point, lower = ci[2, ], upper = ci[1, ])

  return(ep)
}
